datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -786,20 +886,42 @@ impl Stream for SMJStream {
                         }
                     } else {
                         self.freeze_all()?;
-                        if !self.output_record_batches.is_empty() {
+                        if !self.output_record_batches.batches.is_empty() {
                             let record_batch = self.output_record_batch_and_reset()?;
+                            let record_batch = if !(self.filter.is_some()
+                                && matches!(
+                                    self.join_type,
+                                    JoinType::Left | JoinType::LeftSemi
+                                )) {
+                                record_batch
+                            } else {
+                                RecordBatch::new_empty(Arc::clone(&self.schema))
+                            };
+
                             return Poll::Ready(Some(Ok(record_batch)));
                         }
                         return Poll::Pending;
                     }
                 }
                 SMJState::Exhausted => {
                     self.freeze_all()?;
-                    if !self.output_record_batches.is_empty() {
-                        let record_batch = self.output_record_batch_and_reset()?;
-                        return Poll::Ready(Some(Ok(record_batch)));
+
+                    if !self.output_record_batches.batches.is_empty() {
+                        if self.filter.is_some()
+                            && matches!(
+                                self.join_type,
+                                JoinType::Left | JoinType::LeftSemi
```

Comment:
there is multiple if statements of this kind just because we moved only 2 types of join